File: sort/bubblesort/bubblesort.c
#include <stdio.h>

void bubblesort(int a[], int size) {
    int i, j;
    int changed;
    int tmp;
    int step = 0;
    for (i = 0; i < size - 1; i ++) {
        changed = 0;
	// size - 1 - i, is already sorted
        for (j = 0; j < size - 1 - i; j ++) {
        // for (j = 0; j < size - 1; j ++) {
            
            if (a[j] > a[j + 1]) {
                tmp = a[j + 1];
                a[j + 1] = a[j];
                a[j] = tmp;
                changed = 1;
            }
	    step ++;
        }
        
        if (changed == 0) {
            break;
        }
        
    }

    printf("step: %d\n", step);
    
}


void main() {
    int a[11] = { 9, 8, 7, 6, 5, 4,3, 3,  2, 1, 0};
    bubblesort(a, sizeof(a)/ sizeof(a[0]));
    int i;
    for (i = 0; i < sizeof(a)/ sizeof(a[0]); i ++) {
        printf("%d ", a[i]);
    }
    
    printf("\n");
    
}

File: sort/bubblesort/bubblesort.go
package main

import (
    "fmt"
)

func bubbleSort(data []int) {
    length := len(data)
    for i := 0; i < length; i ++ {
    // 每次循环把最小或最大的移动到最右边，所以只需要排列前面的就可以
    // 外循环i表示排列好了几个，所以j只需要排列到length - i - 1
    // 因为下面交换的时候是j + 1可以达到最后一个，所以需要length -i -1
        for j := 0; j < length - i - 1 ; j ++ {
            if data[j] < data[j + 1] {
                tmp := data[j]
                data[j] = data[j + 1]
                data[j + 1] = tmp
            }
        }
    }
}

func main() {
    data := []int{9, 8, 7, 6, 6, 4, 5, 3, 3, 1, 0, 2}
    bubbleSort(data)
    for _, v := range data {
        fmt.Println(v)
    }
}